scantool/diag_l2_kwp6227.c

```diff
@@ -0,0 +1,286 @@
+ * diagnostics on the 1996-1998 Volvo 850, S40, C70, S70, V70, XC70, V90 and
+ * possibly other models.
+ *
+ * The message headers are similar, but not identical, to KWP2000.
+ * In KWP2000, the length value in the header represents the number of
+ * data bytes only in the message; in KWP6227, it also includes the trailing
+ * checksum byte -- that is, the length value is 1 greater in KWP6227 than it
+ * would be in KWP2000.
+ *
+ * This driver currently works only with ELM327 interfaces.
+ *
+ */
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include "diag.h"
+#include "diag_tty.h"
+#include "diag_err.h"
+#include "diag_os.h"
+#include "diag_l1.h"
+#include "diag_l2.h"
+#include "diag_l2_kwp6227.h"
+
+/* replace a byte's msb with a parity bit */
+enum eo { PARITY_EVEN=0, PARITY_ODD=1 };
+static uint8_t
+with_parity(uint8_t c, enum eo eo)
+{
+	uint8_t p = eo;
+	int i;
+
+	for (i = 0; i < 7; i++) {
+		p ^= c; p <<= 1;
+	}
+
+	return((c&0x7f)|(p&0x80));
+}
+
+static int
+dl2p_6227_send(struct diag_l2_conn *d_l2_conn, struct diag_msg *msg)
+{
+	int rv;
+	uint8_t buf[3 + 14 + 1];
+	struct diag_l2_kwp6227 *dp;
+
+	dp = (struct diag_l2_kwp6227 *)d_l2_conn->diag_l2_proto_data;
+
+	if (msg->len < 1 || msg->len > 14)
+		return diag_iseterr(DIAG_ERR_BADLEN);
+
+	buf[0] = 0x80 + msg->len + 1;
+	buf[1] = msg->dest ? msg->dest : dp->dstaddr;
+	buf[2] = msg->src ? msg->src : dp->srcaddr;
+
+	memcpy(&buf[3], msg->data, msg->len);
+
+	diag_os_millisleep(d_l2_conn->diag_l2_p3min);
+
+	rv = diag_l1_send(d_l2_conn->diag_link->l2_dl0d, NULL, buf,
+		msg->len + 3, d_l2_conn->diag_l2_p4min);
+
+	return rv?diag_iseterr(rv):0;
+}
+
+static int
+dl2p_6227_recv(struct diag_l2_conn *d_l2_conn, unsigned int timeout,
+	void (*callback)(void *handle, struct diag_msg *msg),
+	void *handle)
+{
+	int rv;
+	uint8_t buf[3 + 14 + 1];
+	struct diag_msg *msg;
+
+	rv = diag_l1_recv(d_l2_conn->diag_link->l2_dl0d, NULL, buf,
+		sizeof(buf), timeout + 100);
+	if (rv < 0)
+		return rv;
+
+	msg = diag_allocmsg((size_t)(rv - 4));
+	if (msg == NULL)
+		return diag_iseterr(DIAG_ERR_NOMEM);
+	memcpy(msg->data, &buf[3], (size_t)(rv - 4));
+	msg->rxtime = diag_os_chronoms(0);
+	msg->src = buf[2];
+	msg->dest = buf[1];
+	msg->fmt = DIAG_FMT_FRAMED;
+
+	if (callback)
+		callback(handle, msg);
+
+	diag_freemsg(msg);
+
+	return 0;
+}
+
+static void
+dl2p_6227_request_callback(void *handle, struct diag_msg *in)
+{
+	struct diag_msg **out = (struct diag_msg **)handle;
+	*out = diag_dupsinglemsg(in);
+}
+
+static struct diag_msg *
+dl2p_6227_request(struct diag_l2_conn *d_l2_conn, struct diag_msg *msg,
+                int *errval)
+{
+	int rv;
+	struct diag_msg *rmsg;
+
+	rv = diag_l2_send(d_l2_conn, msg);
+	if (rv < 0) {
+		*errval = rv;
+		return NULL;
+	}
+
+	rv = dl2p_6227_recv(d_l2_conn, 1000, dl2p_6227_request_callback, &rmsg);
+	if (rv < 0) {
+		*errval = rv;
+		return rmsg;
+	}
+	if (rmsg == NULL)
+		*errval = DIAG_ERR_NOMEM;
+
+	return rmsg;
```

Comment:
to be totally rigorous, the normal, error-free exit with "return rmsg" should have an explicit " *errval = 0;"  ; I've had this trigger some compiler warnings in some conditions.